File: apps/Pose-Motion/iOS/PoseMotion/Pipeline/PersonCropper.swift
import CoreGraphics

/// Person box → pose-model crop rect: pad, then expand to the model's 3:4 (w:h)
/// aspect around the box center, clamped to the frame.
enum PersonCropper {
    static func cropRect(for personBox: CGRect, frameSize: CGSize) -> CGRect {
        let aspect = CGFloat(AppConfig.poseInputWidth) / CGFloat(AppConfig.poseInputHeight)

        let boxW = personBox.width * frameSize.width * AppConfig.cropPadding
        let boxH = personBox.height * frameSize.height * AppConfig.cropPadding
        let cx = (personBox.midX) * frameSize.width
        let cy = (personBox.midY) * frameSize.height

        // Fit to 3:4 by growing the short side.
        var w = boxW
        var h = boxH
        if w / h > aspect {
            h = w / aspect
        } else {
            w = h * aspect
        }

        var rect = CGRect(x: cx - w / 2, y: cy - h / 2, width: w, height: h)

        // Clamp: shift inside the frame first, then intersect as a last resort.
        if rect.minX < 0 { rect.origin.x = 0 }
        if rect.minY < 0 { rect.origin.y = 0 }
        if rect.maxX > frameSize.width { rect.origin.x = frameSize.width - rect.width }
        if rect.maxY > frameSize.height { rect.origin.y = frameSize.height - rect.height }
        rect = rect.intersection(CGRect(origin: .zero, size: frameSize))
        return rect
    }
}

File: apps/Pose-Motion/Android/app/src/main/java/ai/zetic/demo/posemotion/pipeline/PersonCropper.kt
package ai.zetic.demo.posemotion.pipeline

import ai.zetic.demo.posemotion.AppConfig
import android.graphics.RectF

/**
 * Person box → pose-model crop rect: pad, then expand to the model's 3:4 (w:h)
 * aspect around the box center, clamped to the frame. Returns frame-pixel coords.
 */
object PersonCropper {
    fun cropRect(personBox: RectF, frameW: Float, frameH: Float): RectF {
        val aspect = AppConfig.POSE_INPUT_WIDTH.toFloat() / AppConfig.POSE_INPUT_HEIGHT

        val boxW = personBox.width() * frameW * AppConfig.CROP_PADDING
        val boxH = personBox.height() * frameH * AppConfig.CROP_PADDING
        val cx = personBox.centerX() * frameW
        val cy = personBox.centerY() * frameH

        // Fit to 3:4 by growing the short side.
        var w = boxW
        var h = boxH
        if (w / h > aspect) h = w / aspect else w = h * aspect

        var left = cx - w / 2
        var top = cy - h / 2

        // Clamp: shift inside the frame first, then intersect as a last resort.
        if (left < 0f) left = 0f
        if (top < 0f) top = 0f
        if (left + w > frameW) left = frameW - w
        if (top + h > frameH) top = frameH - h
        val rect = RectF(left, top, left + w, top + h)
        rect.intersect(0f, 0f, frameW, frameH)
        return rect
    }
}
